### ISSUE TITLE
[HOLD for payment 2022-12-20] [$500] Allow native app releases to take staging env variables into consideration

### ISSUE THREAD

**Issue by mountiny** [Expensify/App#11561]
### Problem

In this issue https://github.com/Expensify/App/issues/10163, we have implemented a setting toggle which will aim all the API commands to staging or production API (if in Staging App). However, this works only in desktop or web, but not in native mobile apps as the live logs shown.

The reason it does not work is that we always build the App only once for deploy. We just relabel staging release for the production one and so when building the staging App we use the production env variables causing this issue

#### Actual behaviour

1. On native mobile App (iOS and Android), the staging API toggle is turned on
2. The API requests are still made to production API endpoints

#### Expected behaviour

1. On native mobile App (iOS and Android), the staging API toggle is turned on
2. The API requests will be made to the staging endpoints same as on web app version

### Why is it important

This will allow us to test staging changes with staging API at the same time.

### Solution

Find a way how we can use the environment variables in native app so we can distinguish between staging and production to send the API requests to desired API endpoints.

**Comment by mountiny** [comment]
I realized we have this [Staging badge or dev badge](https://github.com/Expensify/App/blob/6748e8a6c5363a626490e436c84ecd53df7266c6/src/components/EnvironmentBadge.js#L1-L30) even in the native apps so we must have some way to figure that out and the [withEnvironment HOC](https://github.com/Expensify/App/blob/6748e8a6c5363a626490e436c84ecd53df7266c6/src/components/withEnvironment.js#L1-L60)

I am not sure if we can use those in the API toggle but would be worth discovering.

**Comment by MitchExpensify** [comment]
[Upwork Job](https://www.upwork.com/jobs/~01f3701354e5318181)

**Comment by parasharrajat** [comment]
No proposals yet

**Comment by smrutiparida** [comment]
> We just relabel staging release for the production one and so when building the staging App we use the production env variables causing this issue

It looks difficult to me to imagine something while we continue to release the andorid/ios app by relabeling the staging release.

What I want to propose instead is to use https://github.com/luggit/react-native-config npm and issue the build commands like

$ ENVFILE=.env.staging react-native run-ios                  //for staging and

$ ENVFILE=.env.production react-native run-ios           //for production

I can try the same. Request you comment @mountiny

**Comment by parasharrajat** [comment]
@smrutiparida We already do that. Please check the code and let us know if you feel something is off.

**Comment by parasharrajat** [comment]
@mountiny I think the issue details are not very clear. I don't understand the expected behavior. Can you please improve these?

**Comment by smrutiparida** [comment]
My second solution above can help the android code read a version no from a firebase remote config. Based on the version stored in remote, the API_URL can be configured dynamically within in the code. By doing so, the android build can be tested against staging URL for all versions higher than the remote version. It can also be released to PROD by only changing the label. @mountiny is that the expected thing ?

**Comment by mountiny** [comment]
@parasharrajat @smrutiparida I have added some lines to the issue body. The problem is that the native apps disregard the API endpoint settings, they always send requests to the production API even if the staging API toggle in settings in turned on.

I am unfortunately not that familiar with firebase remote config so I am not sure. Is there any way we could use this approach: https://github.com/Expensify/App/issues/11561#issuecomment-1266689289 to achieve this natively?

**Comment by parasharrajat** [comment]
Thanks for adding that details. It is a lot more clear to me now.

**Comment by smrutiparida** [comment]
@mountiny What I have understood so far is, we rename the label of a staging build to release to PROD. This is the reason API_URL present in the ENV values are as PROD_URL as there are no way presently for an iOS/Android android build to read the env values dynamically based on dev/taging/prod post the creation of a build.

@parasharrajat Which setting you mean ?

**Comment by mountiny** [comment]
> What I have understood so far is, we rename the label of a staging build to release to PROD. This is the reason API_URL present in the ENV values are as PROD_URL as there are no way presently for an iOS/Android android build to read the env values dynamically based on dev/taging/prod post the creation of a build.

Yes we take the code in main and build it with prod settings and release as Staging version. What used to be staging version is then relabelled to production so no changes to the code.

**Comment by parasharrajat** [comment]
There are three env files with the name `.env.<environemnt>`. so IMO, `.env.staging` is used while building the staging app.

**Comment by smrutiparida** [comment]
@parasharrajat @mountiny shall we work on it or hold on to this issue ?

**Comment by mountiny** [comment]
> There are three env files with the name .env.<environemnt>. so IMO, .env.staging is used while building the staging app.

I suspect they might be used for web and desktop but what I understood about the native deploy pipeline the app does not get the staging env variables the same way as on web, hence this issue.

@smrutiparida no, this issue is still persistent and would be great to fix it

**Comment by smrutiparida** [comment]
> 2. Another way can be reading variable from firebase remote config
> 
> ```
> // remote config with production version if same found in the build then 
> // app points to PROD_URL for everything else STAGING_URL is set for testing. 
> // We can improvise this logic.
> 
> const checkAppVersion = remoteConfig().getValue('app_production_version');
> 
> if (checkAppVersion.asString() === getBuildVersion()) {
>   setEnvironmentAsProduction();
> }
> else 
> {
>   setEnvironmentAsStaging();
> }
> ```
> 
> https://github.com/invertase/react-native-firebase/tree/main/packages/remote-config
> 

Hi

I was thinking to try this with Firebase config. Are we looking for a proposal beyond above.

**Comment by parasharrajat** [comment]
Any proposal is accepted that fixes the issue.

**Comment by mountiny** [comment]
Correct, but ideally we prefer to solve this without adding new packages.

**Comment by smrutiparida** [comment]
**** Proposal

1. First, I moved the of the URLs based on security and environment value lib/Environment files

<img width="893" alt="Screen Shot 2022-10-20 at 15 07 30" src="https://user-images.githubusercontent.com/5945400/196915350-94d0b3ed-4713-4db4-98b9-ed842b36a457.png">

2. Second, basic implementation of returning the URLs based on environment value
<img width="813" alt="Screen Shot 2022-10-20 at 15 12 12" src="https://user-images.githubusercontent.com/5945400/196915794-6c719fc3-675d-4ab6-9458-bf5b726ed9b5.png">

3. BetaChecker.isBetaBuild() shall set the environment for ios and android based on existing code and return the proper environment value 
We just moved the shouldUseStagingServer check to index.js and index.native.js as part of keeping the code in one place.
<img width="791" alt="Screen Shot 2022-10-20 at 15 12 25" src="https://user-images.githubusercontent.com/5945400/196916030-d12502e6-6e82-4f5a-9165-bea165660803.png">

4. Do the same for desktop and web in index.js
<img width="789" alt="Screen Shot 2022-10-20 at 15 13 06" src="https://user-images.githubusercontent.com/5945400/196916958-2054b64d-7c4a-4ee7-82b8-6e7210384941.png">

Please review @mountiny @parasharrajat

**Comment by parasharrajat** [comment]
All config should be in `Config.js` file. The environment will only give us env info.

**Comment by smrutiparida** [comment]
What we are solving here is,

In the fastlane config i.e ./Fastfile we have the below config. 
```
desc "Generate a new local APK"
  lane :build do
    ENV["ENVFILE"]=".env.production"

    gradle(
      project_dir: './android',
      task: 'assemble',
      build_type: 'Release',
    )
  end
```

Because of which a build always calls the production API URL configured in the .env.production. And since we use the same build for testing and just relabels it as mentioned in description of this bug, quoted below.

> The reason it does not work is that we always build the App only once for deploy. We just relabel staging release for the production one and so when building the staging App we use the production env variables causing this issue

With this change, even  a Fastfile production configured build by comparing between the package version and the version uploaded to play store presently, shall be able to figure out if the build is the PROD one or STAGING one. Knowing which it shall use the staging API URL for staging build and PROD API URL for prod build

**Comment by smrutiparida** [comment]
@mountiny could you please suggest what to do next ?

**Comment by mountiny** [comment]
@smrutiparida I am not sure on this one, I am not that familiar with the deploy flow so dont want to agree with changes which might not be intended. I will ask someone who might be more familiar

**Comment by AndrewGable** [comment]
I agree the root cause is because we always use `.env.production` in fastlane builds. I also agree that using our existing `BetaChecker.isBetaBuild()` is best here. We should be able to come up with a clean solution that uses `BetaChecker.isBetaBuild()` and `.env.staging`, right? 

I think @smrutiparida is close to the ideal solution, but maybe there is some additional code cleanup @parasharrajat or @mountiny can suggest..

**Comment by parasharrajat** [comment]
Feedback on the proposal https://github.com/Expensify/App/issues/11561#issuecomment-1285351812: 
1. Config variables should be in Config.js file. you can move the logic to dynamically calculate it in that file but all of those will **only** be exported from this file.

**Comment by parasharrajat** [comment]
Waiting for proposals...

**Comment by smrutiparida** [comment]
>We should be able to come up with a clean solution that uses BetaChecker.isBetaBuild() and .env.staging, right?

The present issue is to use the Staging API_URL in Staging build. Which is achieved by BetaChecker.isBetaBuild() check. Using .env.staging for staging build shall need change in Fastlane process which is a larger discussion. Hence I think that change can be kept out of scope of this issue and we can create another task for that. 

> Feedback on the proposal [#11561 (comment)](https://github.com/Expensify/App/issues/11561#issuecomment-1285351812):
> 
> 1. Config variables should be in Config.js file. you can move the logic to dynamically calculate it in that file but all of those will **only** be exported from this file.

Please name the config variables we should move. 
The ENVIRONMENT_URLS constant is already present Environment.js and we have introduced 2 new Constants for API_URL named SECURE_ENVIRONMENT_URLS and UNSECURE_ENVIRONMENT_URLS. Since those 2 constants are local to Environment.js. moving that to Config.js may not be needed.

**Comment by flodnv** [comment]
@AndrewGable @parasharrajat thoughts?

**Comment by parasharrajat** [comment]
I am still confused about the root cause. I will give it another try in some time.

**Comment by mountiny** [comment]
@smrutiparida I agree that using the entire staging config is not what we exactly ask for in this issue, but it would be great to get some sense if that is ultimately what we want to achieve (I think it would make sense as at the moment it is confusing and easy to just assume Staging versions use the env.staging variables).

I think it should be fine to do the necessary changes in this issue to allow the Staging API toggle in Android and iOS native apps too.

@smrutiparida To follow the contributor process can you please post a proposal here how that would be achieved. As @AndrewGable mentioned, you are close to he solution, can you please clean it up based on the feedback and we can review it again? Thank you very much!

**Comment by smrutiparida** [comment]
> it is confusing and easy to just assume Staging versions use the env.staging variables 

@mountiny The solution won't be that straight forward. The present Fastlane process does not even read the .env.staging values. how shall we read those values. We would need a totally different approach for that. 

>I think it should be fine to do the necessary changes in this issue to allow the Staging API toggle in Android and iOS native apps too.

In the PR #10690 we introduced STAGING_EXPENSIFY_URL and STAGING_SECURE_EXPENSIFY_URL. This change shall help you read these 2 values in an android/ios staging build by using isBetaCheckBuild() function. Proposal below.

> We should be able to come up with a clean solution that uses BetaChecker.isBetaBuild() and .env.staging, right?

@AndrewGable .env.staging file is no where read by the Fastlane build. A clean solution here should be to totally get away with the .env files and use firebase config to read the environment values. I have mentioned about that solution 
here https://github.com/Expensify/App/issues/11561#issuecomment-1291493698

<img width="947" alt="Screen Shot 2022-11-07 at 13 39 45" src="https://user-images.githubusercontent.com/5945400/200258536-263948bd-32a6-4fd2-b770-e31623151923.png">

**Comment by smrutiparida** [comment]
### Proposal

Use additional check of Environment.getEnvironment() along with IS_IN_STAGING configuration to known the environment values for setting API URL. 
BetaChecker.isBetaBuild() shall set the environment for ios and android based on existing code and returns the proper environment value 

<img width="885" alt="Screen Shot 2022-11-07 at 14 05 12" src="https://user-images.githubusercontent.com/5945400/200263456-19a918cb-22a9-45d0-a3f6-ecd00aa8c672.png">

**Comment by MitchExpensify** [comment]
Created a new [Upwork Job](https://www.upwork.com/jobs/~018a04b2e237644730) (and invited @parasharrajat for eventual C+ payment) as the last one expired.

**Comment by mountiny** [comment]
Looks promising to me @parasharrajat what are your thoughts?

**Comment by mountiny** [comment]
@smrutiparida just wanted to confirm, we cannot use the isBetaChecker for web/desktop too? Just wondering it would be nice to only have one method to distinguish that env variables on all platforms

**Comment by smrutiparida** [comment]
@mountiny You're right. We can remove CONFIG.IS_IN_STAGING  and just use the Environment.getEnvironment to know both. The BetaChecker code support it for web/desktop and android/ios. 

However some additional testing shall be needed as the web/desktop change of this functionality introduced the variable in the config in #10690. Can you see that previous PR and confirm if I shall remove it or not.

<img width="687" alt="Screen Shot 2022-11-07 at 21 51 27" src="https://user-images.githubusercontent.com/5945400/200365080-d3a98e8c-882e-4324-a955-40a7f1cb0176.png">

Once you confirm, the change shall look like this

<img width="959" alt="Screen Shot 2022-11-07 at 22 07 44" src="https://user-images.githubusercontent.com/5945400/200364991-cbe5ea7e-1943-4c28-a935-f561a2606428.png">

**Comment by mountiny** [comment]
Yeah I think that looks fine to me, I would just wait for Rajat to confirm.

The change itself makes sense to me however, I cannot predict without testing this will work compared to just using the CONFIG

**Comment by parasharrajat** [comment]
My mind is blocked atm. I will check back in sometime.

**Comment by mountiny** [comment]
No problem, make sure to rest. I think we can go ahead with @smrutiparida [proposal](https://github.com/Expensify/App/issues/11561#issuecomment-1305882992), I hope you have tested this locally. We can try to polish this in the PR.

Please, feel free to raise the PR

**Comment by mountiny** [comment]
Seems fair to me, can we update the price to 500 or handle this by the extra bonus? @MitchExpensify

**Comment by MitchExpensify** [comment]
Updated the price as I'm less likely to forget that way :D

**Comment by mountiny** [comment]
This is in review, not overdue.

**Comment by mountiny** [comment]
We are having some tough time deciding next steps as the PR review and testing showed some drawbacks of this solution we are not sure are worth it.

I have suggested bringing the discussion Slack to see what others might think about the side-effects.

**Comment by mountiny** [comment]
Working on a new PR, hopefully that will work just fine for us. @smrutiparida you will be compensated for your work here even if we dont go with your solution.

**Comment by mountiny** [comment]
This is not done yet, original PR has been reverted and the fix is being worked on.

**Comment by mountiny** [comment]
The PR has been updated and is ready for a review.

**Comment by parasharrajat** [comment]
I will take a look.

**Comment by trjExpensify** [comment]
@mountiny we're just waiting on a deploy here, right?

**Comment by mountiny** [comment]
Yeah this is in staging and I have QAed and it worked at least from what I saw 🎉

**Comment by trjExpensify** [comment]
Noice! ❤️

**Comment by MitchExpensify** [comment]
Reassigning the Bug label as I'm heading ooo and this will need a new CM!

**Comment by mountiny** [comment]
@kadiealexander Coming from [this comment](https://github.com/Expensify/App/issues/11561#issuecomment-1308976504) we should have a payment off:
- 500 to @smrutiparida for their efforts to fix this

**Comment by kadiealexander** [comment]
Thanks @mountiny!!!! Is payment due now?

**Comment by mountiny** [comment]
I would say yes, this has been PR which was kind of untraditional in terms of the process. The internal PR should already be in production.

**Comment by kadiealexander** [comment]
Thanks @mountiny! New job post is [here](https://www.upwork.com/jobs/~012d55f943ee1df801), I've already sent hire contracts to @parasharrajat and @smrutiparida, please let me know once you've accepted and I'll issue payment.

**Comment by mountiny** [comment]
> ETA: @mountiny correct me if I'm wrong, but the problem behind this issue doesn't look like it can be added to TestFlight for regression testing? Just confirming before I take action on the checklist above.

Yeah I dont think think Applause could test the native apps which is the main issue here
